### PR TITLE
BreakpointWidget: Allow breakpoints to be toggled between enable/disable

### PR DESCRIPTION
**Pull request by InusualZ**
This PR add a way to disable/enable breakpoints at runtime, by adding a context menu that let you toggle between the two state.
Also, added a `Go to` action, that let you go to breakpoint address, this action is only added when the breakpoint is a instruction breakpoint.

Context Menu

I removed the `itemClicked` listener since it was replaced with the context menu `Go to` action.

### CODE REVIEW

**Comment by sepalani** [comment]
@dolphin-emu-bot rebuild

**Review by sepalani — CHANGES_REQUESTED** [review]

File: Source/Core/Core/PowerPC/BreakPoints.cpp [hunk 1]
```diff
@@ -174,15 +196,13 @@ void MemChecks::AddFromStrings(const TMemChecksStr& mc_strings)
     std::stringstream ss;
```
[the remark this answers is not in the document]
Comment (InusualZ) [reply Source/Core/Core/PowerPC/BreakPoints.cpp]: Nice catch!

File: Source/Core/Core/PowerPC/BreakPoints.cpp [hunk 2]
```diff
@@ -175,13 +175,9 @@ MemChecks::TMemChecksStr MemChecks::GetStrings() const
...
   {
     std::ostringstream ss;
-    ss << std::hex << mc.start_address << " "
-       << mc.end_address << " "
-       << (mc.is_enabled ? "n" : "")
-       << (mc.is_break_on_read ? "r" : "")
-       << (mc.is_break_on_write ? "w" : "")
-       << (mc.log_on_hit ? "l" : "")
-       << (mc.break_on_hit ? "b" : "");
+    ss << std::hex << mc.start_address << " " << mc.end_address << " " << (mc.is_enabled ? "n" : "")
```
[the remark this answers is not in the document]
Comment (InusualZ) [reply Source/Core/Core/PowerPC/BreakPoints.cpp:180]: First time that have amended a commit other than the previous one. What a rush! xD

File: Source/Core/Core/PowerPC/BreakPoints.cpp [hunk 3]
```diff
@@ -155,10 +175,12 @@ MemChecks::TMemChecksStr MemChecks::GetStrings() const
   for (const TMemCheck& mc : m_mem_checks)
   {
     std::ostringstream ss;
-    ss << std::hex << mc.start_address;
-    ss << " " << (mc.is_ranged ? mc.end_address : mc.start_address) << " "
-       << (mc.is_ranged ? "n" : "") << (mc.is_break_on_read ? "r" : "")
-       << (mc.is_break_on_write ? "w" : "") << (mc.log_on_hit ? "l" : "")
+    ss << std::hex << mc.start_address << " "
```
[the remark this answers is not in the document]
Comment (InusualZ) [reply Source/Core/Core/PowerPC/BreakPoints.cpp]: Done

**Comment by sepalani** [comment]
@dolphin-emu-bot rebuild

**Review by sepalani — COMMENTED** [review]
Do you mind adding `ss.imbue` in `BreakPoints::GetStrings` and `BreakPoints::AddFromStrings` as well? Otherwise, LGTM.

File: Source/Core/Core/PowerPC/BreakPoints.cpp [hunk 4]
```diff
@@ -110,6 +116,20 @@ void BreakPoints::Add(u32 address, bool temp, bool break_on_hit, bool log_on_hit
   JitInterface::InvalidateICache(address, 4, true);
 }
 
+bool BreakPoints::ToggleBreakPoint(u32 address)
+{
+  for (auto& bp : m_breakpoints)
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/BreakPoints.cpp:121]: std::find_if can be used here. Ditto for the memchecks equivalent function

File: Source/Core/DolphinQt/Debugger/BreakpointWidget.cpp [hunk 5]
```diff
@@ -310,6 +308,60 @@ void BreakpointWidget::OnSave()
...
+
+  auto* menu = new QMenu(this);
+
+  const auto& inst_breakpoints = PowerPC::breakpoints.GetBreakPoints();
+  const auto& bp_iter =
+      std::find_if(inst_breakpoints.begin(), inst_breakpoints.end(),
+                   [bp_address](const auto& bp) { return bp.address == bp_address; });
+  if (bp_iter != inst_breakpoints.end())
+  {
+    menu->addAction(tr(bp_iter->is_enabled ? "Disable" : "Enable"), [this, &bp_address]() {
```
Comment (lioncash) [inline Source/Core/DolphinQt/Debugger/BreakpointWidget.cpp:332]:
```suggestion
    menu->addAction(bp_iter->is_enabled ? tr("Disable") : tr("Enable"), [this, &bp_address] {
```

File: Source/Core/DolphinQt/Debugger/BreakpointWidget.cpp [hunk 6]
```diff
@@ -310,6 +308,60 @@ void BreakpointWidget::OnSave()
...
+                     [bp_address](const auto& bp) { return bp.start_address == bp_address; });
+
+    if (mb_iter == memory_breakpoints.end())
+    {
+      return;
+    }
+
+    is_memory_breakpoint = true;
+
+    menu->addAction(tr(mb_iter->is_enabled ? "Disable" : "Enable"), [this, &bp_address]() {
```
Comment (lioncash) [inline Source/Core/DolphinQt/Debugger/BreakpointWidget.cpp:352]:
```suggestion
    menu->addAction(mb_iter->is_enabled ? tr("Disable") : tr("Enable"), [this, &bp_address] {
```

File: Source/Core/DolphinQt/Debugger/BreakpointWidget.cpp [hunk 7]
```diff
@@ -310,6 +308,60 @@ void BreakpointWidget::OnSave()
...
+    menu->addAction(tr(bp_iter->is_enabled ? "Disable" : "Enable"), [this, &bp_address]() {
+      PowerPC::breakpoints.ToggleBreakPoint(bp_address);
+      Update();
+    });
+  }
+  else
+  {
+    // It should be a memory breakpoint
+    const auto& memory_breakpoints = PowerPC::memchecks.GetMemChecks();
+    const auto& mb_iter =
```
Comment (lioncash) [inline Source/Core/DolphinQt/Debugger/BreakpointWidget.cpp:341]:
```suggestion
    const auto mb_iter =
```

File: Source/Core/DolphinQt/Debugger/BreakpointWidget.cpp [hunk 8]
```diff
@@ -310,6 +308,60 @@ void BreakpointWidget::OnSave()
...
+
+  const auto& selected_item = selected_items.constFirst();
+  const auto bp_address = static_cast<u32>(selected_item->data(Qt::UserRole).toUInt());
+
+  auto is_memory_breakpoint = false;
+
+  auto* menu = new QMenu(this);
+
+  const auto& inst_breakpoints = PowerPC::breakpoints.GetBreakPoints();
+  const auto& bp_iter =
```
Comment (lioncash) [inline Source/Core/DolphinQt/Debugger/BreakpointWidget.cpp:327]:
```suggestion
  const auto bp_iter =
```

**Comment by sepalani** [comment]
@dolphin-emu-bot rebuild

**Review by sepalani — APPROVED** [review]
LGTM

**Comment by JMC47** [comment]
Tested and appears to work.

**Comment by JMC47** [comment]
I don't know how strict we are with things, but there's a typo in the commit - Allow to disable/enable from the Breakpoin**t**Widget

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Review by leoetlino — APPROVED** [review]

**Comment by trevor403** [comment]
Fixes comma separated hex values caused by locale
e.g. 
https://github.com/dolphin-emu/dolphin/pull/9456#pullrequestreview-573535894

**Pull request merged**